### ISSUE TITLE
Qwen3 architecture scaling experiments

### ISSUE THREAD

**Issue by dianaonutu** [OpenEuroLLM/Taskboard#196]
## Compare Qwen3 architecture with the Opensci one from the [original scaling experiments](https://github.com/OpenEuroLLM/dense_english_scaling_results) of Niccolo. 

The model architectures were taken from [here](https://docs.google.com/spreadsheets/d/129mvSQ2K3m_hLaH2zk8R2Zr1-RsFyhTphXyN_ErSWFs/edit?gid=2047296841#gid=2047296841). The tokenizer used in all experiments is **GPT-NEOX 20B**.

### **Opensci** 
Spreadsheet tab: Ref 50k v5 (REFERENCE procedure)

|                | 50M | 130M | 300M | 600M | 1B  |
|-----------|------|------|------|-----|------|
| Model Dim      | 480  | 576  | 896  | 1280| 1536 |
| FFN Factor     | 4   | 4    | 4    | 4   | 4    |
| Heads          | 6   | 9    | 14   | 20  | 24   |
| Layers         | 12  | 18   | 20   | 20  | 24   |
| FFN Dim        | 1536| 2304 | 3584 | 5120| 6144 |
| GLU            | TRUE| TRUE | TRUE | TRUE| TRUE |
| Bias           | TRUE| TRUE | TRUE | TRUE| TRUE |
| QK Norm        | TRUE| TRUE | TRUE | TRUE| TRUE |
| Tie Embedding  | TRUE| TRUE | TRUE | TRUE| TRUE |

### **Qwen3** 
Spreadsheet tab: Qwen3 + GPT-OSS tokenizer
|                | 150M | 300M | 600M | 1.1B |
|----------------|------|------|------|------|
| Model Dim      | 512  | 768  | 1024 | 1536 |
| Head Size      | 128  | 128  | 128  | 128  |
| FFN Factor     | 3    | 3    | 3    | 3    |
| Intermediate   | 1536 | 2304 | 3072 | 4608 |
| Heads          | 8    | 8    | 16   | 16   |
| Layers         | 16   | 22   | 28   | 28   |
| QueryGroups    | 4    | 4    | 8    | 8    |
| GLU            | TRUE | TRUE | TRUE | TRUE |
| Bias           | FALSE| FALSE| FALSE| FALSE|
| QK Norm        | TRUE | TRUE | TRUE | TRUE |
| Tie Embedding  | TRUE | TRUE | TRUE | TRUE |

**Comment by dianaonutu** [comment]
- The Qwen3 experiments were run with `oellm-autoexp` on Leonardo. Configurations can be found [here](https://github.com/OpenEuroLLM/oellm-autoexp/tree/exp_diana/config/experiments/diana/qwen3_gpt_neox)

- All plots, except the wandb ones, are based on the loss calculated on an independent validation set: `/leonardo_work/OELLM_prod2026/datasets/Nemotron-cc-2024-HQ-LUMI-sample-valid/high-all`

- The original dense scaling experiments were run on LUMI with `autoexperiment` tool.

- [OLD] Wandb project: [qwen3_experiments](https://wandb.ai/openeurollm-project/qwen3_experiments). Found later that these experiments don't employ GQA.

- [WIP] [Wandb report](https://wandb.ai/openeurollm-project/qwen3_experiments/reports/Open-sci-vs-Qwen-style-architecture--VmlldzoxNjI2NTUyNA) for comparison of runs: Qwen vs Opensci

**Comment by dianaonutu** [comment]
To Dos:

- [x] add autorestart rule to oellm-autoexp for the common errors encountered during experiments: https://github.com/OpenEuroLLM/oellm-autoexp/commit/c0a0af23525b4c960bc05335a620f3fbbe162884
- [x] evaluate loss on independent validation set based on Niccolo's configs
- [ ] evaluate downstream performance
- [ ] add configurations for testing throughput and mbsz

**Comment by dianaonutu** [comment]
## Initial Qwen3 experiments
Qwen3 experiments with model sizes from 150M up to 1B parameters trained on token budgets up to 300BT took: 
- roughly 1 week, including multiple crashed runs, restarts, and maintenance on Leonardo
- approximately 46.5K GPU hours on Leonardo calculated based on elapsed time from `sacct`


These experiments didn't sweep LR and GBSZ, but rather used the (near-)optimal values from the original dense scaling experiments from [here](https://github.com/OpenEuroLLM/dense_english_scaling_results/tree/main/fig/loss/grid). The runs used the entire Nemotron-CC high dataset as input for training data, with the validation set automatically derived by Megatron, whereas the Opensci style runs used a ~300 subsample. 

These results are based on Opensci models (results from Niccolo's runs on LUMI) and initial Qwen3 experiments. Despite intending to run the models with GQA, we've later found that this was not the case.

Pretty large difference in validation loss on small model scales. For example, Qwen600M performs similarly to Opensci300M. This difference decreases with model scale.
<img width="1490" height="490" alt="Image" src="https://github.com/user-attachments/assets/009bb76b-f554-400d-8aed-cbfc0071d170" />

>Note that the lines don't represent the exact same run. Depending on the token budget, different hyperparameters were used. However, the hyperparametes of both model types match at a certain model size and token budget.

Oellm-autoexp [configurations](https://github.com/OpenEuroLLM/oellm-autoexp/blob/exp_diana/config/experiments/diana/qwen3_gpt_neox/eval_qwen3_all.yaml) used for calculating the loss on the independent validation dataset are presented below.

**Comment by dianaonutu** [comment]
## We've taken a step back and run some small experiments with Opensci models on 130M up to 50BT with LR 0.002 and GBSZ128.

All models are from now on trained using the exact same data subsample for training (and validation during training).

**[EDIT 15/04/26]**: I ran a second replication experiment, which exactly reproduced Niccolo's results. This also indicates a small variation in performance across runs with identical configurations.

### 1. Replication experiment

- oellm-autoexp [config ](https://github.com/OpenEuroLLM/oellm-autoexp/blob/exp_diana/config/experiments/diana/replication/dense_130M_50BT_lumi_subsample.yaml)
- oellm-autoexp validation set evaluation [config](https://github.com/OpenEuroLLM/oellm-autoexp/blob/exp_diana/config/experiments/diana/replication/eval_opensci_dense_130M_50BT.yaml)
- wandb [report](https://wandb.ai/openeurollm-project/dense_scaling/reports/Baseline-Runs-Check--VmlldzoxNTc5NDAxMQ)

The training and validation loss seem to match. 
<img width="1113" height="381" alt="Image" src="https://github.com/user-attachments/assets/4d905501-0aa2-4ee2-ae5e-52e77d9ce856" />

From the figure below, the validation loss on the independent validation set almost matches.
**[EDIT 15/04/26]** I ran a second replication experiment, which exactly reproduced Niccolo's results. This also indicates a small variation in performance across runs with identical configurations. Overall, the replication experiments succeeded.

<img width="1200" height="400" alt="Image" src="https://github.com/user-attachments/assets/9b029fca-71c9-4548-8c98-a7d685855fde" />

### 2. Opensci no bias
- oellm-autoexp [config](https://github.com/OpenEuroLLM/oellm-autoexp/blob/exp_diana/config/experiments/diana/testing/dense_130M_50BT_lumi_subsample_biases.yaml)
- oellm-autoexp validation set evaluation [config ](https://github.com/OpenEuroLLM/oellm-autoexp/blob/exp_diana/config/experiments/diana/testing/eval_130M_50BT_biases.yaml)

**[EDIT 15/04/26]**: Added the second replication experiment and adjusted conclusions.
With bias or without bias, the validation loss on the independent validation set matches; it seems that bias has a negligible effect at this scale (130M-50BT). 
> Megatron outputs the same number of non-embedding parameters in both cases (bias/no bias), as this is rounded; thus, the token axis is still used.

<img width="1200" height="400" alt="Image" src="https://github.com/user-attachments/assets/e0e6d842-c49b-4663-8b87-ac3f000597f6" />

### 3. Opensci with GQA
- oellm-autoexp [config ](https://github.com/OpenEuroLLM/oellm-autoexp/blob/exp_diana/config/experiments/diana/testing/dense_130M_50BT_lumi_subsample_gqa.yaml)
- oellm-autoexp validation set evaluation [config](https://github.com/OpenEuroLLM/oellm-autoexp/blob/exp_diana/config/experiments/diana/testing/eval_130M_50BT_gqa.yaml)

To use GQA, we reduced the number of attention heads from 9 to 8 and set the number of query groups to 4. By using GQA, we slightly reduce the number of parameters. This modification decreases the parameter count, from 0.10B non-embedding parameters in the original model to 0.09B with GQA.

**[EDIT 15/04/26]**: Added the second replication experiment and adjusted conclusions.
The validation loss of the Opensci model with GQA may seem slightly higher than that of the original model. However, the two models have a slightly different parameter count. Overall, both loss curves follow a similar trajectory. This suggests that the performance difference is likely due to the reduced parameter count rather than a problem with the model architecture.

<img width="1200" height="400" alt="Image" src="https://github.com/user-attachments/assets/1d2be304-235b-44f4-9cd5-350e9313af64" />

Plots are produced in this [notebook](https://github.com/OpenEuroLLM/oellm-autoexp/blob/exp_diana/analysis.ipynb).

**Comment by dianaonutu** [comment]
## Coming back to Qwen3 models: Corrected GQA usage

We've re-run the 150M and 300M Qwen3 models using the architecture provided in the initial tables and actually employed GQA. Also, the runs are trained on the exact same (~300BT) subsample from LUMI as the Opensci runs.

- oellm-autoexp configs for [Qwen 150M](https://github.com/OpenEuroLLM/oellm-autoexp/blob/exp_diana/config/experiments/diana/qwen3_gpt_neox/dense_qwen_150M_gpt_neox.yaml) and [Qwen 300M](https://github.com/OpenEuroLLM/oellm-autoexp/blob/exp_diana/config/experiments/diana/qwen3_gpt_neox/dense_qwen_300M_gpt_neox.yaml)

- oellm-autoexp validation set evaluation [config](https://github.com/OpenEuroLLM/oellm-autoexp/blob/exp_diana/config/experiments/diana/qwen3_gpt_neox/eval_qwen3_all.yaml). A bit of a hacky way of using it by commenting out unnecessary model sizes...

All runs (Qwen and Opensci) are based on optimal hyperparameters (LR and GBSZ) at each models size and token budget. Thus, two points on the same line may represent a run with different LR and GBSZ. We observe that even at matching compute budget, Qwen models lag behind in performance compared to the Opensci ones. However, this can be likely attributed to the model size differences, as Sampo explains it [here](https://mattermost.ufal.mff.cuni.cz/openeurollm/pl/6aggbufabir8jr3sph5oo9pxja).

<img width="1200" height="400" alt="Image" src="https://github.com/user-attachments/assets/581b7136-37a4-4e89-8081-e39030360669" />

Plots are produced in this [notebook](https://github.com/OpenEuroLLM/oellm-autoexp/blob/exp_diana/analysis.ipynb).

**Comment by dianaonutu** [comment]
## Mismatch in early-stage training loss for Qwen3 "150M" between my configurations and [Sampo's configs](https://mattermost.ufal.mff.cuni.cz/openeurollm/pl/mkmkta8e578kfr1zfj96cgqhyc).

Over the past weeks, both Sampo and I ran experiments using a Qwen3-style architecture. Sampo's results can be found [here](https://mattermost.ufal.mff.cuni.cz/openeurollm/pl/6aggbufabir8jr3sph5oo9pxja). We observed a mismatch [here](https://mattermost.ufal.mff.cuni.cz/openeurollm/pl/jpd96taippnib8x88pfymbedjo) in early-stage training loss for the Qwen3 “150M” model: my run showed a spiky loss curve and higher loss compared to Sampo’s. We investigated this matter below.

**tl;dr**: `beta2=0.99` (instead of 0.95) causes the early-stage training instability and `kv_channels=64` (instead of 128) causes the higher loss value likely due to reduced number of parameters.

To investigate this, we compared training configurations and identified differences across model architecture, optimization, and other hyperparameters. The differences in model architecture also vary the total number of parameters.

| |Diana| Sampo |
| ---    | ----| --- |
|group_query_attention| True | False |
|kv_channels|64 | 128| 
|num_query_groups|4|1|
|adam_beta2|   0.99  |       0.95 |
| attention_softmax_in_fp32 | False | True |
|lr|0.002|0.004
|min_lr| 0.00001 |0|
|weight_decay|0.1|0.05|

The early-stage instability is caused by setting adam `beta2=0.99`. This is evident when taking Sampo’s configuration and modifying only this parameter (see figure below). Despite the instability, the loss magnitude remains close to that of the original configuration. 

<img width="1600" height="840" alt="Image" src="https://github.com/user-attachments/assets/8930e894-8521-423f-b1ae-26d94d028c50" />

Updating only `beta2` in my setup removes the spikes and yields a more stable curve (green). However, the loss remains higher than Sampo’s (yellow). 

<img width="1600" height="840" alt="Image" src="https://github.com/user-attachments/assets/396f246f-a376-42ef-b478-fe4dca91d895" />

The most impactful hyperparameter to reduce the loss is `kv_channels=128`, which decreases the loss and brings it close to Sampo’s results.

<img width="2528" height="1328" alt="Image" src="https://github.com/user-attachments/assets/5ee133c2-8e1e-4534-8dd2-e357dbad939c" />

Additional observations:

- Using or not using FP32 for the attention softmax has negligible effect on the loss
- Weight decay also has negligible effect 
- Enabling or disabling GQA (matters for the total parameter count) has a minor impact on the loss value
- Disabling GQA & setting `lr=0.004` (in addition to `beta2=0.95` and `kv_channels=128`) would bring the loss of "my" configuration closest to Sampo's 

<img width="2528" height="1328" alt="Image" src="https://github.com/user-attachments/assets/1b29f553-e506-43cd-9441-f3dc4369fbe6" />

The experiments can be found in the `sanity-checks` [wandb project](https://wandb.ai/openeurollm-project/sanity-checks).
